docs/okeps/okep-5259-no-overlay.md

```diff
@@ -0,0 +1,1104 @@
+  * `routeAdvertisements`: Specifies the name of the RouteAdvertisements CR used
+    to advertise the pod network of this CUDN. This field acts as a
+    discriminator, creating an explicit link between the RouteAdvertisements CR,
+    the network, and the user's intent to prevent misconfiguration.
+
+The `spec` field of a ClusterUserDefinedNetwork CR is immutable. Therefore, the
+overlay configuration cannot be changed after a ClusterUserDefinedNetwork CR is
+created.
+
+```golang
+type OverlayProtocol string
+type SnatOption string
+
+const (
+    OverlayProtocolNone   OverlayProtocol = "None"
+    OverlayProtocolGeneve OverlayProtocol = "Geneve"
+
+    SnatEnable  SnatOption = "Enable"
+    SnatDisable SnatOption = "Disable"
+)
+
+// NoOverlayOptions contains configuration options for networks operating in no-overlay mode.
+type NoOverlayOptions struct {
+    // OutboundSNAT defines the SNAT behavior for outbound traffic from pods.
+    // +kubebuilder:validation:Enum=Enable;Disable
+    // +kubebuilder:default=Disable
+    // +optional
+    OutboundSNAT SnatOption `json:"outboundSNAT,omitempty"`
+    // RouteAdvertisements specifies the name of the RouteAdvertisements CR
+    // used to advertise the pod network of this CUDN.
+    // +optional
+    RouteAdvertisements string `json:"routeAdvertisements,omitempty"`
+}
+
+type NetworkSpec struct {
+    ...
+    // Overlay describes the overlay protocol for east-west traffic.
+    // Allowed values are "None" and "Geneve".
+    // - "None": The network operates in no-overlay mode.
+    // - "Geneve": The network uses Geneve overlay.
+    // Defaults to "Geneve".
+    // +kubebuilder:validation:Enum=None;Geneve
+    // +kubebuilder:default=Geneve
+    // +optional
+    Overlay OverlayProtocol `json:"overlay,omitempty"`
+    // NoOverlayOptions contains configuration for no-overlay mode.
+    // It is required when Overlay is "None".
+    // +optional
+    NoOverlayOptions *NoOverlayOptions `json:"noOverlayOptions,omitempty"`
+}
+```
+
+To prevent a user from setting unsupported no-overlay configurations in CUDN
+accidentally, we will add the following CEL validation rules to the
+`spec.network` field of the ClusterUserDefinedNetwork CRD.
+
+```yaml
+    network:
+      x-kubernetes-validations:
+      - message: "overlay 'None' is only supported for Layer3 topology"
+        rule: 'self.overlay != "None" || self.topology == "Layer3"'
+      - message: "noOverlayOptions is required when overlay is 'None'"
+        rule: 'self.overlay != "None" || has(self.noOverlayOptions)'
+      - message: "noOverlayOptions is only supported for no-overlay networks"
+        rule: 'self.overlay == "None" || !has(self.noOverlayOptions)'
+      - message: "noOverlayOptions.routeAdvertisements is required when overlay is 'None'"
+        rule: 'self.overlay != "None" || (has(self.noOverlayOptions) && has(self.noOverlayOptions.routeAdvertisements) && self.noOverlayOptions.routeAdvertisements != "")'
```

Comment:
Uhm... if we make this `RouteAdvertisements` field required, what value should it have if we want to support the simplest use case we discussed, where the cluster admin only wants to enable no-overlay for all eligible networks between nodes and doesn't want to export pod subnets to the outside? In this use case we're in charge of creating the base FRRConfig and the RA.